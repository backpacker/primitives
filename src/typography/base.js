import React, { useContext } from 'react';
import { Text as RNText } from 'react-native';
import PropTypes from 'prop-types';

import ConfigContext from '../config-context';
import { TextStylePropTypes } from '../prop-types';
import { makeStyle } from '../utils/style';

const Text = (props) => {
  const {
    children,

    uppercase,
    underline,
    strikeout,

    extralight,
    thin,
    light,
    medium,
    semibold,
    bold,
    heavy,
    extraheavy,

    center,

    testID,
    ...rest
  } = props;

  const { theme } = useContext(ConfigContext);
  const { fontFamily, colors, weights } = theme;

  const baseStyle = makeStyle([
    {
      fontFamily,
      color: colors.text,
      textAlign: 'left'
    },
    uppercase && { textTransform: 'uppercase' },
    underline && { textDecorationLine: 'underline' },
    strikeout && { textDecorationLine: 'line-through' },
    center && { textAlign: 'center' },

    extralight && { fontWeight: weights.extralight },
    thin && { fontWeight: weights.thin },
    light && { fontWeight: weights.light },
    medium && { fontWeight: weights.medium },
    semibold && { fontWeight: weights.semibold },
    bold && { fontWeight: weights.bold },
    heavy && { fontWeight: weights.heavy },
    extraheavy && { fontWeight: weights.extraheavy },

    { ...rest }
  ]);

  return (
    <RNText style={baseStyle} testID={testID}>
      {children}
    </RNText>
  );
};

Text.propTypes = {
  ...TextStylePropTypes,
  center: PropTypes.bool,
  uppercase: PropTypes.bool,
  underline: PropTypes.bool,
  strikeout: PropTypes.bool,
  extralight: PropTypes.bool,
  thin: PropTypes.bool,
  light: PropTypes.bool,
  medium: PropTypes.bool,
  semibold: PropTypes.bool,
  bold: PropTypes.bool,
  heavy: PropTypes.bool,
  extraheavy: PropTypes.bool
};

Text.defaultProps = {
  center: false,
  uppercase: false,
  underline: false,
  strikeout: false,
  extralight: false,
  thin: false,
  light: false,
  medium: false,
  semibold: false,
  bold: false,
  heavy: false,
  extraheavy: false
};

export default Text;
