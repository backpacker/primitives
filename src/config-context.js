import React, { useContext, useMemo, useState } from 'react';
import PropTypes from 'prop-types';

import baseConfig from './config';

const ConfigContext = React.createContext(undefined);

function ConfigProvider({ config: customConfig, defaultTheme, children }) {
  const [activeTheme, setActiveTheme] = useState(defaultTheme);

  const theme = useMemo(() => {
    const config = { ...baseConfig, ...customConfig };

    const baseTheme = baseConfig.default;

    const defaultSpacerSize =
      config[activeTheme]?.defaultSpacerSize || baseTheme.defaultSpacerSize;
    const spacerUnit = config[activeTheme]?.spacerUnit || baseTheme.spacerUnit;

    const spacing = (size = defaultSpacerSize) => size * spacerUnit;

    return {
      ...config[activeTheme],
      spacing
    };
  }, [activeTheme, customConfig]);

  const value = {
    theme,
    setTheme: setActiveTheme
  };

  return (
    <ConfigContext.Provider value={value}>{children}</ConfigContext.Provider>
  );
}

ConfigProvider.propTypes = {
  config: PropTypes.shape({}),
  defaultTheme: PropTypes.string
};

ConfigProvider.defaultProps = {
  config: {},
  defaultTheme: 'default'
};

const withTheme = (Component) => (props) =>
  (
    <ConfigContext.Consumer>
      {(value) => <Component {...props} {...value} />}
    </ConfigContext.Consumer>
  );

function useTheme() {
  const ctx = useContext(ConfigContext);
  if (!ctx) {
    throw new Error(`useTheme() must be called within ConfigContext`);
  }
  return ctx;
}

export { ConfigProvider, withTheme, useTheme };
export default ConfigContext;
