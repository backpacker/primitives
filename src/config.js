export const weights = {
  extralight: '100',
  thin: '200',
  light: '300',
  normal: '400',
  medium: '500',
  semibold: '600',
  bold: '700',
  heavy: '800',
  extraheavy: '900'
};

const defaultTheme = {
  fontSize: {
    h1: 32,
    h2: 24,
    h3: 20,
    h4: 16,
    h5: 14,
    h6: 12,
    h7: 10,
    h8: 8
  },
  fontWeight: weights.normal,
  fontFamily: undefined,
  weights,

  spacerUnit: 8,
  defaultSpacerSize: 2,
  spacer: {
    xs: 10,
    s: 14,
    m: 18,
    l: 22,
    xl: 26
  },

  colors: {
    brand: '#28D9AA',
    background: '#FFF',
    text: '#000'
  },

  isDark: false
};

const darkTheme = {
  ...defaultTheme,
  colors: {
    brand: '#28D9AA',
    background: '#000',
    text: '#FFF'
  },
  isDark: true
};

export { defaultTheme };

export default {
  activeTheme: 'default',
  default: defaultTheme,
  dark: darkTheme
};
