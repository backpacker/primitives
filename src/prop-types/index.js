/**
 * Style PropTypes
 */

import LayoutStylePropTypes from './layout-style';
import TouchableStylePropTypes from './touchable-style';
import TextStylePropTypes from './text-style';
import ModalPropTypes from './modal';

export {
  LayoutStylePropTypes,
  TouchableStylePropTypes,
  TextStylePropTypes,
  ModalPropTypes
};
