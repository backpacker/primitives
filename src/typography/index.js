import React, { useContext } from 'react';

import Text from './base';
import ConfigContext from '../config-context';
import config from '../config';

const makeHeadingComponent = ({ key }) => {
  return ({ children, ...rest }) => {
    const { theme } = useContext(ConfigContext);
    const { fontSize } = theme;

    return (
      <Text fontSize={fontSize[key]} {...rest}>
        {children}
      </Text>
    );
  };
};

const [H1, H2, H3, H4, H5, H6, H7, H8] = Object.entries(
  config.default.fontSize
).map(([key]) => makeHeadingComponent({ key }));

export { Text, H1, H2, H3, H4, H5, H6, H7, H8 };
