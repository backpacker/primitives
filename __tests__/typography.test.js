/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';

import { ConfigProvider } from '../src/config-context';
import { Text, H1 } from '../src/typography';

describe('<Text>', () => {
  it('should render Alan Turing; default formatting', () => {
    const tree = renderer
      .create(
        <ConfigProvider>
          <Text>Alan Turing</Text>
        </ConfigProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render Alan Turing; bold', () => {
    const tree = renderer
      .create(
        <ConfigProvider>
          <Text bold>Alan Turing</Text>
        </ConfigProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render Alan Turing; uppercase underline strikeout semibold center', () => {
    const tree = renderer
      .create(
        <ConfigProvider>
          <Text uppercase underline strikeout semibold center>
            Alan Turing
          </Text>
        </ConfigProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render Alan Turing; thin fontSize={18} color={#FFF}', () => {
    const tree = renderer
      .create(
        <ConfigProvider>
          <Text thin fontSize={18} color='#FFF'>
            Alan Turing
          </Text>
        </ConfigProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('<H1>', () => {
  it('should render Alan Turing', () => {
    const tree = renderer
      .create(
        <ConfigProvider>
          <H1>Alan Turing</H1>
        </ConfigProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
