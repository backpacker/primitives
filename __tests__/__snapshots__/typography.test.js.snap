// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text> should render Alan Turing; bold 1`] = `
<Text
  style={
    Object {
      "color": "#000",
      "fontFamily": undefined,
      "fontWeight": "700",
      "textAlign": "left",
    }
  }
>
  Alan Turing
</Text>
`;

exports[`<Text> should render Alan Turing; default formatting 1`] = `
<Text
  style={
    Object {
      "color": "#000",
      "fontFamily": undefined,
      "textAlign": "left",
    }
  }
>
  Alan Turing
</Text>
`;

exports[`<Text> should render Alan Turing; thin fontSize={18} color={#FFF} 1`] = `
<Text
  style={
    Object {
      "color": "#FFF",
      "fontFamily": undefined,
      "fontSize": 18,
      "fontWeight": "200",
      "textAlign": "left",
    }
  }
>
  Alan Turing
</Text>
`;

exports[`<Text> should render Alan Turing; uppercase underline strikeout semibold center 1`] = `
<Text
  style={
    Object {
      "color": "#000",
      "fontFamily": undefined,
      "fontWeight": "600",
      "textAlign": "center",
      "textDecorationLine": "line-through",
      "textTransform": "uppercase",
    }
  }
>
  Alan Turing
</Text>
`;

exports[`<H1> should render Alan Turing 1`] = `
<Text
  style={
    Object {
      "color": "#000",
      "fontFamily": undefined,
      "fontSize": 32,
      "textAlign": "left",
    }
  }
>
  Alan Turing
</Text>
`;
